eclair-core/src/main/scala/fr/acinq/eclair/io/Switchboard.scala

```diff
@@ -114,8 +126,64 @@ class Switchboard(nodeParams: NodeParams, authenticator: ActorRef, watcher: Acto
   override val supervisorStrategy = OneForOneStrategy(loggingEnabled = true) { case _ => SupervisorStrategy.Resume }
 }
 
-object Switchboard {
+object Switchboard extends Logging {
 
   def props(nodeParams: NodeParams, authenticator: ActorRef, watcher: ActorRef, router: ActorRef, relayer: ActorRef, wallet: EclairWallet) = Props(new Switchboard(nodeParams, authenticator, watcher, router, relayer, wallet))
 
+  /**
+    * If we have stopped eclair while it was forwarding HTLCs, it is possible that we are in a state were an incoming HTLC
+    * was committed by both sides, but we didn't have time to send and/or sign the corresponding HTLC to the downstream node.
+    *
+    * In that case, if we do nothing, the incoming HTLC will eventually expire and we won't lose money, but the channel will
+    * get closed, which is a major inconvenience.
+    *
+    * This check will detect this and will allow us to fast-fail HTLCs and thus preserve channels.
+    *
+    * @param channels
+    * @return
+    */
+  def checkBrokenHtlcsLink(channels: Seq[HasCommitments]): Seq[UpdateAddHtlc] = {
+
+    // We are interested in incoming HTLCs, that have been *cross-signed*. They signed it first, so the HTLC will first
+    // appear in our commitment tx, and later on in their commitment when we subsequently sign it.
+    // That's why we need to look in *their* commitment with direction=OUT.
+    val htlcs_in = channels
+      .flatMap(_.commitments.remoteCommit.spec.htlcs)
+      .filter(_.direction == OUT)
+      .map(_.add)
+
+    // Here we do it differently because we need the origin information.
+    val relayed_out = channels
+      .flatMap(_.commitments.originChannels.values)
+      .collect { case r: Relayed => r }
+      .toSet
+
+    val htlcs_broken = htlcs_in.filterNot(htlc_in => relayed_out.exists(r => r.originChannelId == htlc_in.channelId && r.originHtlcId == htlc_in.id))
```

Comment:
What happens for direct payments that we've received ? there is no corresponding `Relayed` origin so they will be considered broken ?

Reply:
oops, good catch